### ISSUE TITLE
Misconfiguration of prod-preview environment

### ISSUE THREAD

**Issue by ScrewTSW** [openshiftio/openshift.io#2063]
Hello.

The configuration of prod-preview environment causes issues with Bayesian not working.
We have tested the version of rh-che that had the issues on the production environment and our independent dev clusters and minishift solutions.

The issue is only present on che.prod-preview.openshift.io

Could you please look into this issue?
If additional information is needed, feel free to ask :)

original issue: https://github.com/redhat-developer/rh-che/issues/524

![](https://user-images.githubusercontent.com/16451875/35336977-4a1b56a8-011a-11e8-8833-762ac91e07e8.png)
![](https://user-images.githubusercontent.com/16451875/35336991-510fc282-011a-11e8-873d-4bfa2a908f03.png)

**Comment by rootAvish** [comment]
@sivaavkd This should be deprecated to SEV-3, it's not a prod issue.

**Comment by ibuziuk** [comment]
@ScrewTSW @rhopp is it still reproducible for che 6 ?

**Comment by invincibleJai** [comment]
@ibuziuk nope it's not working on che 6 in prod-preview, the reason I feel is env variable is not coming for "RECOMMENDER_API_URL" it's coming only for "RECOMMENDER_API_TOKEN"
`ps -awx |grep RECOMMENDER_API_URL`  
<img width="1221" alt="screen shot 2018-04-24 at 11 19 00 pm" src="https://user-images.githubusercontent.com/5129024/39204840-9a1eeeb0-4816-11e8-8cbf-427a020fa74a.png">

**Comment by sivaavkd** [comment]
@sunix @ibuziuk any comments on this please? cc @invincibleJai

**Comment by l0rd** [comment]
@invincibleJai @sivaavkd it looks like `$RECOMMENDER_API_URL` is not passed on both prod-preview and prod (c.f. the screenshot below, white is osio prod, black is osio prod-preview). 

![image](https://user-images.githubusercontent.com/606959/40620746-338014f6-629a-11e8-82ec-6072cf7cdbb9.png)

So I am afraid that this is not the root cause of bayesian plugin not working on prod-preview.

What we think is that we are using prod-preview users osio tokens against production recommendation server `recommender.api.openshift.io/api/v1/` and the requests are not authorized (I can see "status 401" messages in prod-preview `/home/user/che/ls-bayesian/ls-bayesian.log`).  

Is there a different URL for prod-preview recommmender endpoint that we should use?

**Comment by l0rd** [comment]
Digging a little more I have figured out that when `$RECOMMENDER_API_URL` is not provided ls-bayesian set it to prod URL:

```javascript
this.server_url = process.env.RECOMMENDER_API_URL || "https://recommender.api.openshift.io/api/v1";
```

So @invincibleJai that may be the root cause of the problem after all. But we still need the prod-preview value for `$RECOMMENDER_API_URL`.

**Comment by l0rd** [comment]
ping @invincibleJai @sivaavkd what is the correct value for `RECOMMENDER_API_URL` on prod-preview?

**Comment by miteshvp** [comment]
@l0rd for prod-preview - 
`RECOMMENDER_API_URL = https://recommender.api.prod-preview.openshift.io/api/v1`

**Comment by l0rd** [comment]
Thank you @miteshvp

**Comment by ibuziuk** [comment]
@miteshvp @invincibleJai @sivaavkd I started to doubt that bayesian have ever worked on prod-preview. Basically, what we are doing on che side is simply downloading `0.0.5` release [2] version of ls and running it during workspace startup. Funny thing is that https://github.com/fabric8-analytics/fabric8-analytics-lsp-server/releases/download/0.0.5/ca-lsp-server.tar contains the following line (that I suspect was manually added for 0.0.5 release to make bayesian working on prod osio )

`this.server_url = process.env.RECOMMENDER_API_URL || "https://recommender.api.openshift.io/api/v1";`

Source code [3]  of `0.0.5` does `not` contains hard-coded version of osio prod recommender api url:

`        this.server_url = process.env.RECOMMENDER_API_URL || "api-url-not-available-in-lsp";
`
Basically, this means that:

- che currently works only against `0.0.5` release of bayesian ls and only against prod e.g. `0.0.7` will work neither on prod nor on prod-preview
- `0.0.5` release of bayesian was done manually and not from source code (line with hardcoded recommender api url was added locally and than archive was manually uploaded to GitHub).
- on che side there is currently no mechanism for setting  `RECOMMENDER_API_URL` env var in workspace pod

Could you please comment if my assumptions are correct ?

[1] https://github.com/redhat-developer/rh-che/blob/master/plugins/ls-bayesian-agent/src/main/resources/installers/1.0.0/com.redhat.bayesian.lsp.script.sh#L15
[2] https://github.com/fabric8-analytics/fabric8-analytics-lsp-server/releases/tag/0.0.5
[3] https://github.com/fabric8-analytics/fabric8-analytics-lsp-server/blob/0.0.5/src/server.ts#L176

**Comment by invincibleJai** [comment]
@ibuziuk yes your assumption is correct, LSP is working on prod in che as 0.0.5 has the fallback url for prod i.e in case env is not found. That is the reason it never worked for prod-preview.

`this.server_url = process.env.RECOMMENDER_API_URL || "https://recommender.api.openshift.io/api/v1";`

If we can have env var for `RECOMMENDER_API_URL` for prod and prod-preview then this should be fine as we have for `RECOMMENDER_API_TOKEN`

https://github.com/fabric8-analytics/fabric8-analytics-lsp-server/blob/0.0.5/src/server.ts#L176-L177

And release process in manual for lsp in CHE. We need to think through this as currently lsp is supported for CHE/DevStudio/VSCode as well.

This task is in backlog https://github.com/openshiftio/openshift.io/issues/672

**Comment by ibuziuk** [comment]
@invincibleJai thanks for clarifications, so it turns out that bayesian never worked on prod-preview and it is not really a regression. In order to support it we need to inject required env var to workspace pod. Currently there is no mechanism for setting config env vars into workspace pod (need to implement environment provider [1] and bind [2] it in rh-che assembly). 

Side discussion. Current version of bayesian ls is 0.0.7 and che is still using 0.0.5. Which team should be in charge of / own of the updates of  bayesian ls in osio ? cc: @l0rd @gorkem @slemeur 

[1]https://github.com/garagatyi/che/blob/2c2bd6637c8169018e2b33397f9d738bbf82a0b1/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/provision/env/EnvVarProvider.java#L23

[2] https://github.com/garagatyi/che/blob/25e574a0e28c7486959d90338eded99a9568b7b5/assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java#L162

**Comment by invincibleJai** [comment]
@ibuziuk Analytics own's bayesian ls in OSIO.

**Comment by miteshvp** [comment]
This looks to be work from che side. Removing `team/analytics` label. 
@ibuziuk - I have added `team/che` generic label, but please assign it to respective che team. Thanks.
